onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx_common_int8.h

```diff
@@ -0,0 +1,758 @@
+        low_mask, zero,
+        QuantBZeroPointPtr, true, scale_00, acc0);
+    }
+
+    *SumPtr = hsum_float_8(acc0);
+    if (BiasPtr) {
+      *SumPtr += *BiasPtr;
+    }
+
+    // move to next column
+
+    QuantBDataColPtr += StrideQuantBData;
+    QuantBScaleColPtr += StrideQuantBScale;
+    if constexpr (HasZeroPoint) {
+      QuantBZeroPointColPtr += StrideQuantBZeroPoint;
+    }
+
+    BiasPtr += BiasPtr != nullptr ? 1 : 0;
+    SumPtr += 1;
+  }
+}
+
+using DotQuadFunctionType = __m256 (*)(
+  const __m256i, const __m256i,
+  const __m256i, const __m256i);
+
+template <bool HasZeroPoint, DotQuadFunctionType dot_quad>
+MLAS_FORCEINLINE void
+ComputeDotProducts_BlkBitWidth4_CompInt8_SubBlkLen64_NCols4(
+  size_t BlkLen,
+  const std::byte* QuantARowPtr,
+  const std::byte* QuantBDataColPtr,
+  const float* QuantBScaleColPtr,
+  const std::byte* QuantBZeroPointColPtr,
+  float* SumPtr,
+  size_t CountK,
+  size_t StrideQuantBData,
+  size_t StrideQuantBScale,
+  size_t StrideQuantBZeroPoint,
+  const float* BiasPtr
+)
+{
+  // TODO: make it work with all BlkLens
+  assert(BlkLen >= 64);
+  constexpr size_t SubBlkLen64 = 64;
+  //const __m256i zero = _mm256_setzero_si256();
+  const __m256i low_mask = _mm256_set1_epi8(0xF);
+
+  constexpr size_t BlkBitWidth = 4;
+  constexpr size_t SubBlkStep = MlasQNBitBlkDataSizeInBytes(BlkBitWidth, SubBlkLen64);
+
+  __m256 acc0 = _mm256_setzero_ps(), acc1 = _mm256_setzero_ps(), acc2 = _mm256_setzero_ps(), acc3 = _mm256_setzero_ps();
+
+  const std::byte* ablob = QuantARowPtr;
+  const std::byte* b_blk_data_ptr = QuantBDataColPtr;
+  const float* blk_scale_ptr = QuantBScaleColPtr;
+
+  [[maybe_unused]] size_t QuantBZeroPointIdx = 0;  // track half byte increments with this index instead of a pointer
+  // only used if HasZeroPoint == true
+
+  for (size_t k = 0; k < CountK; k += BlkLen) {
+    size_t ck = std::min(CountK - k, BlkLen);
+
+    const float a_scale = Q8BlkScale(ablob);
+    ablob += sizeof(float);
+
+    float
+      scale_v0 = (*(blk_scale_ptr + StrideQuantBScale * 0)) * a_scale,
+      scale_v1 = (*(blk_scale_ptr + StrideQuantBScale * 1)) * a_scale,
+      scale_v2 = (*(blk_scale_ptr + StrideQuantBScale * 2)) * a_scale,
+      scale_v3 = (*(blk_scale_ptr + StrideQuantBScale * 3)) * a_scale;
+
+    const std::byte* bptr0 = (b_blk_data_ptr + StrideQuantBData * 0);
+    const std::byte* bptr1 = (b_blk_data_ptr + StrideQuantBData * 1);
+    const std::byte* bptr2 = (b_blk_data_ptr + StrideQuantBData * 2);
+    const std::byte* bptr3 = (b_blk_data_ptr + StrideQuantBData * 3);
+
+    uint8_t zp0, zp1, zp2, zp3;
+    if constexpr (HasZeroPoint) {
+      // TODO: this block causes near 30% of the computation.
+      bool is_lower = (QuantBZeroPointIdx & 1) == 0;
+      std::byte zp_packed = QuantBZeroPointColPtr[0 * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+      zp0 = std::to_integer<int8_t>(is_lower ? (zp_packed & std::byte{ 0x0F }) : (zp_packed >> 4));
```

Comment:
does it help if the branching is moved to a higher level? e.g., `if (is_lower) { zp0 = ..., zp1 = ... } else ...`

Reply:
I do not know. The solution proposed by @yufenglee is to factor out scaleB * zp while packing A. Will do this in next PR.